### PR TITLE
waitSignal and waitSignals() evaluate signal parameters

### PR DESCRIPTION
**Pull request by MShekow**
waitSignal and waitSignals() have a callback/callbacks parameter that makes it possible to test that emitted signals have specific parameters set. waitSignals() also has a force_order parameter that allows to configure 3 different levels of strictness regarding the order in which the signals are expected. See https://github.com/pytest-dev/pytest-qt/issues/118

### CODE REVIEW

File: pytestqt/qtbot.py [hunk 1]
```diff
@@ -234,7 +234,7 @@ def stopForInteraction(self):
 
     stop = stopForInteraction
 
-    def waitSignal(self, signal=None, timeout=1000, raising=None):
+    def waitSignal(self, signal=None, timeout=1000, callback=None, raising=None):
```
Comment (nicoddemus) [inline pytestqt/qtbot.py]: I think this parameter should appear at the end, so code doing `waitSignal(signal, 1000, True)` won't break.

Comment (nicoddemus) [inline pytestqt/qtbot.py]: Also I feel `callback` is to generic... perhaps `check_params_callback` would convey the intent a little better?

**Comment by nicoddemus** [comment]
Hey @MShekow thanks a lot for the PR.

I have to leave to work so I can't review it more carefully now, I will try to make a proper review during lunch, or at the end the day the latest.

File: pytestqt/qtbot.py (hunk 1 above)
Comment (MShekow) [inline pytestqt/qtbot.py]: I don't mind this rename.

**Comment by The-Compiler** [comment]
hmm, no idea why the windows builds are broken...

File: pytestqt/qtbot.py [hunk 2]
```diff
@@ -285,14 +287,14 @@ def waitSignal(self, signal=None, timeout=1000, raising=None):
...
-        blocker = SignalBlocker(timeout=timeout, raising=raising)
+        blocker = SignalBlocker(timeout=timeout, raising=raising, callback=callback)
         if signal is not None:
             blocker.connect(signal)
         return blocker
 
     wait_signal = waitSignal  # pep-8 alias
 
-    def waitSignals(self, signals=None, timeout=1000, raising=None):
+    def waitSignals(self, signals=None, timeout=1000, force_order = "none", callbacks = None, raising=None):
```
Comment (The-Compiler) [inline pytestqt/qtbot.py]: nitpick: No spaces around default values

File: pytestqt/qtbot.py [hunk 3]
```diff
@@ -315,10 +317,19 @@ def waitSignals(self, signals=None, timeout=1000, raising=None):
            blocker.wait()
 
         :param list signals:
-            A list of :class:`Signal` objects to wait for. Set to ``None`` to
-            just use timeout.
+            A list of :class:`Signal`s objects to wait for. Set to ``None`` to just use
```
Comment (The-Compiler) [inline pytestqt/qtbot.py]: That added "s" is wrong :wink:

File: pytestqt/qtbot.py (hunk 2 above)
Comment (The-Compiler) [inline pytestqt/qtbot.py]: Also I think I'd rename `force_order` to `order`, and the parameters should be at the end as well.

File: pytestqt/qtbot.py (hunk 1 above)
Comment (The-Compiler) [inline pytestqt/qtbot.py]: `check_params_cb` maybe? I think `_cb` is a common suffix, and `check_params_callback` seems so long :wink:

File: pytestqt/qtbot.py [hunk 4]
```diff
@@ -315,10 +317,19 @@ def waitSignals(self, signals=None, timeout=1000, raising=None):
...
         :param int timeout:
             How many milliseconds to wait before resuming control flow.
+        :param str force_order:
+            "none": no order is enforced
+            "strict": the signals have to be emitted in the provided order (and no intermediate signals may be emitted)
+            "simple": like "strict", but signals may be emitted inbetween the provided ones, e.g. suppose
+            signals=[a, b, c] and the tested object emits the signals a, a, b, a, c, the test will pass with "simple"
+            but not with "strict"
+        :param list callbacks:
+            optional list of Callable functions that evaluate the parameters of the signal. Each element of the list
```
Comment (The-Compiler) [inline pytestqt/qtbot.py]: "Callable functions" sounds weird - either "list of callables" or "list of functions" IMHO.

File: pytestqt/qtbot.py [hunk 5]
```diff
@@ -334,12 +345,18 @@ def waitSignals(self, signals=None, timeout=1000, raising=None):
 
         .. note:: This method is also available as ``wait_signals`` (pep-8 alias)
         """
+        assert force_order in ["none", "simple", "strict"], "force_order has to be set to 'none', 'simple' or 'strict'"
```
Comment (The-Compiler) [inline pytestqt/qtbot.py]: This should be a `if` raising a `ValueError` - assertions are deactivated with `-O` for example, and an `AssertionError` is a weird thing to get for invalid parameters.

File: pytestqt/wait_signal.py [hunk 6]
```diff
@@ -151,48 +152,121 @@ class MultiSignalBlocker(_AbstractSignalBlocker):
...
             _silent_disconnect(signal, slot)
-        self._signals.clear()
-        self._slots.clear()
+        del self._signals_emitted[:]
+        self._signals_map.clear()
+        del self._slots[:]
 
 
 class SignalEmittedSpy(object):
-
```
Comment (The-Compiler) [inline pytestqt/wait_signal.py]: Unrelated whitespace change

File: tests/test_wait_signal.py [hunk 7]
```diff
@@ -335,7 +336,6 @@ class Obj(QtCore.QObject):
 
 
 class TestArgs:
-
```
Comment (The-Compiler) [inline tests/test_wait_signal.py]: Lots of unrelated whitespace changes :wink:

Comment (MShekow) [inline tests/test_wait_signal.py]: These line changes must have happened when I was using PyCharm's **Reformat code** functionality, which cleans up files very nicely. I'm not familiar with the general rule for this project. Normally, a project should define an open source tool that can be executed on a python module and it will format it according to rules set up by the project team. Is such a mechanism in place?

File: pytestqt/qtbot.py (hunk 4 above)
Comment (MShekow) [inline pytestqt/qtbot.py]: I like "list of callables", as "callable" is a Python term.

File: pytestqt/qtbot.py (hunk 2 above)
Comment (MShekow) [inline pytestqt/qtbot.py]: I'm OK with renaming the parameter to `order` and placing it at the end. Spaces around the default value of the parameter can be removed.

**Comment by MShekow** [comment]
As I've never done PRs, I'd like to know whether the general procedure is this one?:
You suggest changes to the code, I add them to my fork, and once you're satisfied, you merge the PR?

File: tests/test_wait_signal.py (hunk 7 above)
Comment (The-Compiler) [inline tests/test_wait_signal.py]: I don't really know what @nicoddemus uses - but either way I personally think doing cleanup changes as part of other PRs is problematic (when using git blame, when having to revert something, etc.). I think @nicoddemus is a bit more relaxed than I in that regard though :laughing:

**Comment by The-Compiler** [comment]
Right - if you push new changes to your fork, the PR will update automatically. Then we review it again, you update it if needed, etc., until it gets merged.

**Comment by The-Compiler** [comment]
FWIW I get the same AppVeyor issue [for qutebrowser as well](https://ci.appveyor.com/project/The-Compiler/qutebrowser/build/master-3152/job/21wndp808hofslqx). I'll investigate tomorrow.

File: pytestqt/qtbot.py [hunk 8]
```diff
@@ -315,10 +317,19 @@ def waitSignals(self, signals=None, timeout=1000, raising=None):
...
 
         :param list signals:
-            A list of :class:`Signal` objects to wait for. Set to ``None`` to
-            just use timeout.
+            A list of :class:`Signal`s objects to wait for. Set to ``None`` to just use
+            timeout.
         :param int timeout:
             How many milliseconds to wait before resuming control flow.
+        :param str force_order:
+            "none": no order is enforced
```
Comment (nicoddemus) [inline pytestqt/qtbot.py]: Please format those as a list, like in [here](https://github.com/pytest-dev/pytest-qt/blob/master/pytestqt/qtbot.py#L109). The results are [worth it](http://pytest-qt.readthedocs.io/en/1.11.0/reference.html#pytestqt.qtbot.QtBot.mouseRelease) IMO.

File: pytestqt/qtbot.py (hunk 4 above)
Comment (nicoddemus) [inline pytestqt/qtbot.py]: It is not very clear from the docs here what the callbacks are supposed to do with the signals they receive.

File: pytestqt/qtbot.py [hunk 9]
```diff
@@ -315,10 +317,19 @@ def waitSignals(self, signals=None, timeout=1000, raising=None):
...
+            "strict": the signals have to be emitted in the provided order (and no intermediate signals may be emitted)
+            "simple": like "strict", but signals may be emitted inbetween the provided ones, e.g. suppose
+            signals=[a, b, c] and the tested object emits the signals a, a, b, a, c, the test will pass with "simple"
+            but not with "strict"
+        :param list callbacks:
+            optional list of Callable functions that evaluate the parameters of the signal. Each element of the list
+            can be a callable or None. Make sure that the number of callbacks matches the number of signals.
```
Comment (nicoddemus) [inline pytestqt/qtbot.py]: > Make sure that the number of callbacks matches the number of signals.

I suggest make it explicit about the `ValueError`, something like: 

``` 
If the number of callbacks doesn't match the number of signals ``ValueError`` will be raised.
```

File: tests/test_wait_signal.py (hunk 7 above)
Comment (nicoddemus) [inline tests/test_wait_signal.py]: I usually don't mind some whitespace changes in the middle of the PR if they don't get in the way of reviewing, so I think the PR as regarding that. :grin:

File: tests/test_wait_signal.py [hunk 10]
```diff
@@ -371,8 +371,269 @@ def test_connected_signal(self, qtbot, signaller):
         assert blocker.args == ['foo', 2342]
 
 
-class TestAssertNotEmitted:
+def test_signal_identity(signaller):
+    """
```
Comment (nicoddemus) [inline tests/test_wait_signal.py]: Excellent docs! :+1:

File: pytestqt/wait_signal.py [hunk 11]
```diff
@@ -123,6 +122,9 @@ def _quit_loop_by_signal(self, *args):
         """
         quits the event loop and marks that we finished because of a signal.
         """
+        if self.callback:
+            if not self.callback(*args):
+                return  # parameter check did not pass
```
Comment (nicoddemus) [inline pytestqt/wait_signal.py]: I wonder if we should store the received args, so we can provide a better error message? 

Consider:

```python
def is_disconnected(text):
    return text == 'END'

with qtbot.waitSignal(conn.text_received, callback=is_disconnected):
    conn.request_disconnect()   
```

`conn.text_received` is emitted with `['WAIT_CLIENTS', 'PROCESSING']` and eventually times out.

In this situation all the user will see is a `TimeoutError` and won't have any hints at what might have gone wrong. I suppose the user would then have to add `print` statements at `is_disconnected` to try to figure out the problem.

I think if we store the received arguments when there is a callback, we can reuse that later in the exception message:

```
TimeoutError:  `text_received` emitted with ['WAIT_CLIENTS', 'PROCESSING'] 
but did not satisfy the `is_disconnected` callback.
```

That idea also implies that the same arguments should be available for inspection later in case `raising` is `False`, probably as a list of `*args`: 

```python
blocker.all_args == ['WAIT_CLIENTS', 'PROCESSING']
```

**Comment by nicoddemus** [comment]
@MShekow first let me say thank you for your extremely well written PR!

I reviewed it partially, but I'm afraid is very late and I must get some sleep now. I will continue reviewing it tomorrow.

File: pytestqt/qtbot.py (hunk 4 above)
Comment (MShekow) [inline pytestqt/qtbot.py]: So, here is a suggestion:
```
:param list check_param_callbacks: list of callables that compare the provided signal parameter(s) to some expected parameter. Each callable has to match the signature of the corresponding signal in ``signals`` (just like a slot function would) and return ``True`` if parameters match, ``False`` otherwise. Instead of a specific callable, ``None`` can be provided, to disable parameter checking for the corresponding signal. If the number of callbacks doesn't match the number of signals ``ValueError`` will be raised.
```

File: pytestqt/wait_signal.py (hunk 11 above)
Comment (MShekow) [inline pytestqt/wait_signal.py]: Sounds good.
Another question is: should this also be implemented for `MultiSignalBlocker`? It would generally be interesting to get a string-dump (in human readable form) of all recorded signals and their parameters, including the points of time where things went wrong (e.g. in case of _strict_ order when the order was violated). Similarly, if `raising == False`, `blocker.recorded_signals` (or something similar) could contain a list of recorded signals together with their parameters.

**Comment by The-Compiler** [comment]
@nicoddemus can you rerun AppVeyor now that #142 is in? (Also, it'd be nice if this plugin was under the pytestbot account on AppVeyor too..)

**Comment by nicoddemus** [comment]
Don't we have to merge with master to get your changes into this pr anyway? This would cause a new build to be triggered anyway. 

About the pytestbot user, I agree, will see how to do it.

**Comment by The-Compiler** [comment]
As far as I know, PRs on AppVeyor/Travis are always built against the latest master (which is also why they don't build when the PR is unmergeable)

**Comment by nicoddemus** [comment]
> As far as I know, PRs on AppVeyor/Travis are always built against the latest master (which is also why they don't build when the PR is unmergeable)

Oh really didn't notice that, but it makes sense! :wink: 

Just triggered a build, let's see how it goes. I'll look later how to move ownership of the project to `pytestbot`.

**Comment by nicoddemus** [comment]
It broke only the `docs` environment now. :grin:

**Comment by MShekow** [comment]
Hi. Are there any other remarks? I'll be going on vacation for the rest of July soon. Before that, I'll have time to fix all the rather "cosmetic" issues ;). There were 2 larger issues mentioned which I cannot complete before that:
1) Argument-capturing for `SignalBlocker` (`all_args`) and `MultiSignalBlocker` (recording the signals and their args in order)
2) More descriptive error message in `SignalTimeoutError`
The 2 points together are a couple of hours of work, given that I'd need to write tests for these new features, too.

**Comment by nicoddemus** [comment]
Hi @MShekow, sorry for the delay.

If you are going on vacation soon, I would rather we merge it sooner and work on improving the error message in `2.1`. @The-Compiler what do you think?

**Comment by nicoddemus** [comment]
@MShekow if you could go ahead and fix the cosmetic issues it would be great.

**Comment by The-Compiler** [comment]
Sorry for the late answer - I was busy with Europython and a big mail-backlog after that, so I postponed everything which was a bit more work :laughing: 

@nicoddemus I'm guessing @MShekow is in holidays by now - what cosmetic issues do you mean? I can see "docstring of pytestqt.qtbot.QtBot.waitSignal:36: WARNING: Inline emphasis start-string without end-string." on Appveyor, but it seems they fixed the other issues in 2e214195ee2872fa8fe5508b07f3cb4a9a310422.

The PR looks good otherwise (though I don't really understand the order handling and tests 100% yet :laughing:).

I'm wondering if we should make `order="strict"` the default though, as we're already breaking stuff with v2.0 anyways.

**Comment by nicoddemus** [comment]
I will try to merge this tonight together with #130 and make the 2.0 release then.

**Comment by The-Compiler** [comment]
@nicoddemus What do you think about my comment regarding the default ordering above?

**Comment by nicoddemus** [comment]
> I'm wondering if we should make order="strict" the default though, as we're already breaking stuff with v2.0 anyways.

Oh forgot to comment: I agree with using `strict` as default,  but this is a new feature so we won't be breaking anything, right?

**Comment by The-Compiler** [comment]
Nope - `MultiSignalBlocker` already existed before and didn't care in which order the signals were emitted. If we turn on the order checking by default now, that *is* a breaking change.

**Comment by nicoddemus** [comment]
Oh right, sorry.

Hmm not sure TBH, it seems strict ordering only makes sense for signals coming from the same object. For multiple signals from different objects like the [example code in the docs](http://pytest-qt.readthedocs.io/en/1.11.0/signals.html#waitsignals) it doesn't make sense and is even error prone.

**Comment by The-Compiler** [comment]
Hmm, true. I reviewed my usage of `waitSignals` and it has many instances where the order is fixed but I don't actually care (e.g. a `Question` object in a prompt emitting `answered`, `completed` and `answered_no`), and some instances where I care (e.g. `[reply.metaDataChanged, reply.readyRead, reply.finished]`). Let's keep it.

**Comment by nicoddemus** [comment]
Merged. Thanks @MShekow again for the PR!

**Comment by The-Compiler** [comment]
So... I'm a bit late, but I just realised `check_params_cb` and `order` aren't mentioned in the [signal docs](http://pytest-qt.readthedocs.io/en/2.0.0/signals.html) at all, and `order` isn't even in the changelog.

**Comment by nicoddemus** [comment]
Thanks, completely missed that. Good thing the documentation is live, so it can be fixed on `master`. Created #147 for it. :grin:
